src/HealthChecks/HealthChecks/src/DefaultHealthCheckService.cs

```diff
@@ -18,36 +19,83 @@ namespace Microsoft.Extensions.Diagnostics.HealthChecks;
+        ValidateRegistrations(_healthCheckServiceOptions.Value.Registrations);
+
+        _healthReportEntriesQueue = new ConcurrentQueue<KeyValuePair<string, HealthReportEntry>>();
+
+        _periodHealthChecksMap = (from r in _healthCheckServiceOptions.Value.Registrations
+                                  where ByPredicate(r, _healthCheckPublisherOptions.Value.Predicate) // Apply publisher predicate
+                                  where r.Period != Timeout.InfiniteTimeSpan // Skip HCs with no individual period
+                                  group r by r.Period).ToDictionary<IGrouping<TimeSpan, HealthCheckRegistration>, TimeSpan, List<HealthCheckRegistration>>(g => g.Key, g => g.ToList());
+
+        _healthCheckRegistrationTimers = _periodHealthChecksMap.AsParallel().Select(t =>
+            NonCapturingTimer.Create(
+                async (state) =>
+                {
+                    var entries = await RunChecksAsync(t.Value).ConfigureAwait(false);
+                    foreach (var entry in entries)
+                    {
+                        _healthReportEntriesQueue.Enqueue(entry);
+                    }
+                },
+                null,
+                dueTime: healthCheckPublisherOptions.Value.Delay,
+                period: t.Key)).ToList();
     }
-    public override async Task<HealthReport> CheckHealthAsync(
-        Func<HealthCheckRegistration, bool>? predicate,
-        CancellationToken cancellationToken = default)
+
+    public override async Task<HealthReport> CheckHealthAsync(Func<HealthCheckRegistration, bool>? predicate, CancellationToken cancellationToken = default)
     {
-        var registrations = _options.Value.Registrations;
-        if (predicate != null)
-        {
-            registrations = registrations.Where(predicate).ToArray();
-        }
+        // TODO: if we decide for the current design, where HCs with default periodicity are initiated during CheckHealthAsync,
+        // individual-periodicity HCs cannot use the predicate passed in this method (those are initiated in the cctor of DefaultHealthCheckService)
 
         var totalTime = ValueStopwatch.StartNew();
         Log.HealthCheckProcessingBegin(_logger);
 
+        var registrations = _healthCheckServiceOptions.Value.Registrations.Where(r => ByPredicate(r, predicate)).ToList();
+
+        // Run healthchecks with default/publisher period
+        var healthReportEntriesValues = await RunChecksAsync(registrations, cancellationToken).ConfigureAwait(false);
+
+        // Collect health report entries from previously run periodic healthchecks within the default period
+        while (_healthReportEntriesQueue.TryDequeue(out var entry))
+        {
+            healthReportEntriesValues.Add(entry);
+        }
+
+        var healthReportEntriesDictionary = new Dictionary<string, HealthReportEntry>(
+            healthReportEntriesValues.ToDictionary(kv => kv.Key, kv => kv.Value),
```

Comment:
It's possible for one health check to run multiple times between calls to CheckHealthAsync. What happens to ToDictionary for duplicate entries?